core/templates/pages/diagnostic-test-player-page/diagnostic-test.model.ts

```diff
@@ -0,0 +1,256 @@
+  _currentTopicId: string;
+  _eligibleTopicIds: string[];
+  _skippedTopicIds: string[];
+  _passedTopicIds: string[];
+  _failedTopicIds: string[];
+  _topicIdToPrerequisiteTopicIds: TopicIdToRelatedTopicIds;
+  _topicIdToAncestorTopicIds: TopicIdToRelatedTopicIds;
+  _topicIdToSuccessorTopicIds: TopicIdToRelatedTopicIds;
+  setTopicIdToAncestorTopicIds: () => void;
+  setTopicIdToSuccessorTopicIds: () => void;
+  getAncestorsTopicIds: (topicId: string) => string[];
+  getSuccessorTopicIds: (topicId: string) => string[];
+  incrementNumberOfAttemptedQuestions: (incrementByValue: number) => void;
+  recordTopicPassed: () => void;
+  recordTopicFailed: () => void;
+  isTestFinished: () => boolean;
+  getEligibleTopicIds: () => string[];
+  getFailedTopicIds: () => string[];
+  getCurrentTopicId: () => string;
+  setCurrentTopicId: () => void;
+}
+
+export class DiagnosticTestModelData {
+  _totalNumberOfAttemptedQuestions: number;
+  _currentTopicId: string;
+  _eligibleTopicIds: string[];
+  _skippedTopicIds: string[];
+  _passedTopicIds: string[];
+  _failedTopicIds: string[];
+  _topicIdToPrerequisiteTopicIds: TopicIdToRelatedTopicIds;
+  _topicIdToAncestorTopicIds: TopicIdToRelatedTopicIds;
+  _topicIdToSuccessorTopicIds: TopicIdToRelatedTopicIds;
+
+  constructor(topicIdToPrerequisiteTopicIds: TopicIdToRelatedTopicIds) {
+    this._topicIdToPrerequisiteTopicIds = topicIdToPrerequisiteTopicIds;
+    this._totalNumberOfAttemptedQuestions = 0;
+    this._eligibleTopicIds = Object.keys(this._topicIdToPrerequisiteTopicIds);
+    this._skippedTopicIds = [];
+    this._passedTopicIds = [];
+    this._failedTopicIds = [];
+    this._currentTopicId = '';
+    this._topicIdToAncestorTopicIds = {};
+    this._topicIdToSuccessorTopicIds = {};
+    this.setTopicIdToAncestorTopicIds();
+    this.setTopicIdToSuccessorTopicIds();
+  }
+
+  getEligibleTopicIds(): string[] {
+    return this._eligibleTopicIds;
+  }
+
+  getFailedTopicIds(): string[] {
+    return this._failedTopicIds;
+  }
+
+  getPassedTopicIds(): string[] {
+    return this._passedTopicIds;
+  }
+
+  getCurrentTopicId(): string {
+    return this._currentTopicId;
+  }
+
+  getAncestorsTopicIds(topicId: string): string[] {
+    return this._topicIdToAncestorTopicIds[topicId];
+  }
+
+  getSkippedTopicIds(): string[] {
+    return this._skippedTopicIds;
+  }
+
+  getTopicIdToAncestorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToAncestorTopicIds;
+  }
+
+  getTopicIdToSuccessorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToSuccessorTopicIds;
+  }
+
+  getSuccessorTopicIds(topicId: string): string[] {
+    return this._topicIdToSuccessorTopicIds[topicId];
+  }
+
+  incrementNumberOfAttemptedQuestions(incrementByValue: number): void {
+    this._totalNumberOfAttemptedQuestions += incrementByValue;
+  }
+
+  getNumberOfAttemptedQuestions(): number {
+    return this._totalNumberOfAttemptedQuestions;
+  }
+
+  setTopicIdToAncestorTopicIds(): void {
```

Comment:
You're not really just "setting", there's a computation here. So maybe regenerateTopicId...

Reply:
Rephrased